src/java.desktop/macosx/classes/com/apple/laf/AquaComboBoxButton.java

```diff
@@ -154,6 +156,11 @@ public void paintComponent(final Graphics g) {
             } else {
                 painter.state.set(isSquare ? Widget.BUTTON_POP_UP_SQUARE : Widget.BUTTON_POP_UP);
             }
+
+            AccessibleContext ac = this.getAccessibleContext();
+            if (ac != null && (comboBox.getSelectedItem() != null)) {
+                ac.setAccessibleName(comboBox.getSelectedItem().toString());
```

Comment:
Are you sure that the toString() method always provides a valid string, and not a descriptive JSON or some other character set that is useless for a blind user?

Reply:
I tested several times, didn't get such kind of result and as per java docs toString method returns `a string representation of the object.` So, I think it will be a valid string.

Reply:
I tried with the `Integer` object and it works fine. Please suggest if anything else I can test.

Reply:
> I tried with the `Integer` object and it works fine. Please suggest if anything else I can test.

Obviously, it does.

What about a custom object which doesn't override `toString`? Although this shouldn't be a problem either.

What if `JComboBox` uses a non-standard renderer which takes a property from an object? Let's assume, `JComboBox` is populated with `Rectangle` objects, a custom renderer is set which displays the value of the `bottom` field in the combo box. Your code, if I understand it correctly, will default to _full rectangle_ rather than what's really displayed.

How does `JTable` handle this kind of situation? Or `JList`? Swing relies on renderers which makes it capable of displaying an arbitrary object (it uses `toString` by default, if I'm not mistaken), a renderer could provide a custom view of an object in a cell, in a list.

Reply:
I need to check with custom renderer. 

If I understand correctly, with current fix the `getSelectedItem` method will return the `rectangle` object and then calling `toString` method will return rectangle's dimension instead of the displayed text in JComboBox and that may result incorrectly. Right?

Reply:
> If I understand correctly, with current fix the `getSelectedItem` method will return the `rectangle` object and then calling `toString` method will return rectangle's dimension instead of the displayed text in JComboBox and that may result incorrectly. Right?

Yes.

I wonder if other components support this scenario.

Reply:
> What if `JComboBox` uses a non-standard renderer which takes a property from an object? Let's assume, `JComboBox` is populated with `Rectangle` objects, a custom renderer is set which displays the value of the `bottom` field in the combo box. Your code, if I understand it correctly, will default to _full rectangle_ rather than what's really displayed.
> 
You were right. At least in this case the current fix returns _full_rectangle_ dimension in a string form in stead of the text displayed in JComboBox. 
I think it is required to look for alternate fix, probably to get the displayed text from JComboBox and set it as an accessibleName.

 >How does JTable handle this kind of situation? Or JList?

I case of JTable and JList, the accessible context is of type JLabel which returns the accessibleName in string format.

Reply:
> > What if `JComboBox` uses a non-standard renderer which takes a property from an object? Let's assume, `JComboBox` is populated with `Rectangle` objects, a custom renderer is set which displays the value of the `bottom` field in the combo box. Your code, if I understand it correctly, will default to _full rectangle_ rather than what's really displayed.
> 
> You were right. At least in this case the current fix returns _full_rectangle_ dimension in a string form instead of the text displayed in JComboBox.

Thank you for confirming it.

> I had a doubt here if a custom renderer is used to populate the JComboBox, isn't it required to `override` the `toString` method to return useful information from user end?

No, it's not required to override the `toString` method if the programmer doesn't use it.

> If not then I think it is required to look for alternate fix, probably to get the displayed text from JComboBox and set it as an accessibleName.

Could be…

How would you get the text? It should be provided by a renderer…

> > How does JTable handle this kind of situation? Or JList?
> 
> I case of JTable and JList, the accessible context is of type JLabel which returns the accessibleName in string format.

This doesn't really answer the question. Do `JTable` and `JList` correctly display the magnified text when a custom renderer is used? In the case with rectangle and a custom renderer? Do they still display text from `toString`?

If neither `JTable` nor `JList` handle this situation, it's okay to go with `toString` as you implemented.

Reply:
How does `JTable` handle the situation where an image is displayed? You can use SwingSet2 for testing: it displays colored text and images.

Reply:
> This doesn't really answer the question. Do JTable and JList correctly display the magnified text when a custom renderer is used? In the case with rectangle and a custom renderer? Do they still display text from toString?

Checked with JList and it correctly displayed the magnified text for a rectangle and custom renderer.

Reply:
> > This doesn't really answer the question. Do JTable and JList correctly display the magnified text when a custom renderer is used? In the case with rectangle and a custom renderer? Do they still display text from toString?
> 
> Checked with JList and it correctly displayed the magnified text for a rectangle and custom renderer. The `toString` method of `Rectangle` class is not invoked.

It means you have to look into how it's done in JList so that JCheckBox works in a similar way.

Reply:
> It means you have to look into how it's done in JList so that JCheckBox works in a similar way.

For a JList, whenever `setText` is invoked from renderer's `getListCellRendererComponent` method, JLabel's text is set to the value displayed in JList. So, a11y API invokes the `getAccessibleName` method for `AccessibleJLabel` and that returns the correct string.

But in case of JComboBox, a11y API invokes the `getAccessibleName` method for `AccessibleJButton` and that returns an empty string.

To fix this we need to get the current displayed text is JComboBox as the other way around is not correct for custom renderer.

Reply:
> But in case of JComboBox, a11y API invokes the `getAccessibleName` method for `AccessibleJButton` and that returns an empty string.

And this seems to be the problem. The `getAccessibleName` method should be on JComboBox, or its editor, which should return the result of `getAccessibleName` of the renderer component.

Does this work with editable combo box?

Reply:
> And this seems to be the problem. The getAccessibleName method should be on JComboBox, or its editor, which should return the result of getAccessibleName of the renderer component.

Since the a11y APIs are called upon the `AquaComboBoxButton` (subclass of JButton) instance, need to set the text of the button to current displayed text of JComboBox. JComboBox contains `getAccessibleName` method but it is not invoked.

> Does this work with editable combo box?

Yeah, it worked for editable JComboBox. For editable JComboBox, the string is returned by NavigableTextAccessibility and it returns the current text displayed in editor.

Reply:
Accessible name is set by getting the current text displayed in JComboBox (using `renderer getText()` method) in stead of setting it by JComboBox  `getSelectedItem().toString() `method. Tested with custom renderer and it shows the magnified text as displayed in JComboBox.

Reply:
> > And this seems to be the problem. The getAccessibleName method should be on JComboBox, or its editor, which should return the result of getAccessibleName of the renderer component.
> 
> Since the a11y APIs are called upon the `AquaComboBoxButton` (subclass of JButton) instance, need to set the text of the button to current displayed text of JComboBox. JComboBox contains `getAccessibleName` method but it is not invoked.

This is weird. I expect the accessibility API is called on the `JComboBox` in both cases where it's editable or not. Could this be a bug? Should we fix it in a different way?

> > Does this work with editable combo box?
> 
> Yeah, it worked for editable JComboBox. For editable JComboBox, the string is returned by NavigableTextAccessibility and it returns the current text displayed in editor.

That's good.

And it makes me wonder why non-editable combo box behaves so differently.

Reply:
> > This is weird. I expect the accessibility API is called on the JComboBox in both cases where it's editable or not. Could this be a bug? Should we fix it in a different way?
> 
> AFAIK, I don't think that it is a bug because the arrow button in Aqua LAF is an instance of AquaComboBoxButton. So, native side invoked a11y API based on component. Similarly, For non-editable JComboBox, editor is an instance of JTextField and a11y API invoked on NavigableTextAccessibility.

That's the problem: *why* is it called on *the combo button* instead of the combo box itself? It doesn't look right.

In case of editable `JComboBox`, it consists of two components: an editor and the button. I assume the implementation of accessible interfaces gets called on the first child of the combo box. It's about right for the editable component: the editor is the main component of the combo box.

But it doesn't look right to me that the drop-down button is the main component of the combo box if it's non-editable. The combo box should handle the accessibility request itself in this case. It still uses the same renderer which can't be put into editable state. The renderer is never added into containment hierarchy.

Or there could be a *dummy* editor which handles the accessibility.

Reply:
> But it doesn't look right to me that the drop-down button is the main component of the combo box if it's non-editable. The combo box should handle the accessibility request itself in this case. It still uses the same renderer which can't be put into editable state. The renderer is never added into containment hierarchy.

In case of non-editable combobox there is no editor as such and arrowButton is the main component. When native side request a11y information, accessible interface for combo button is returned (as only two accessible component are present. One is arrow button and other is popup). Subsequent accessible APIs are invoked for combo button to get accessible name.

Reply:
> > But it doesn't look right to me that the drop-down button is the main component of the combo box if it's non-editable. The combo box should handle the accessibility request itself in this case. It still uses the same renderer which can't be put into editable state. The renderer is never added into containment hierarchy.
> 
> In case of non-editable combobox there is no editor as such and arrowButton is the main component. When native side request a11y information, accessible interface for combo button is returned (as only two accessible component are present. One is arrow button and other is popup). Subsequent accessible APIs are invoked for combo button to get accessible name.

Do you think it is correct?

I admit refactoring / rewriting accessibility implementation for `JComboBox` is out of scope in this bug. Yet the current way doesn't feel right.

Reply:
> Do you think it is correct?
>
> I admit refactoring / rewriting accessibility implementation for JComboBox is out of scope in this bug. Yet the current way doesn't feel right.

I think so. 
If I am not wrong the non-editable JComboBox doesn't have any editor, so the arrowButton width and height is equal to the combobox width and height. This results in arrowbutton rendering for entire display area of combobox. Due to this we are getting the accessible interface invoked for combobox button.

Reply:
Yes, non-editable combo box doesn't have the editor. Yet the arrowButton is not what displays the text. It is for this reason that I think it is wrong that the button becomes the main component which handles accessibility for combo box if it's not editable. It is not the button that displays the text.

Setting accessible name in `doRendererPaint` doesn't feel right either.

Reply:
As per our discussion, I didn't find other way to set the accessible name.